src/aproxy.py

```diff
@@ -0,0 +1,463 @@
+                ipaddress.ip_network(entry, strict=False)
+                valid_no_proxy.append(entry)
+            except ValueError as exc:
+                # If not an IP/CIDR, check if it's a valid hostname
+                if HOSTNAME_PATTERN.match(entry):
+                    valid_no_proxy.append(entry)
+                else:
+                    raise ValueError(f"invalid no_proxy entry {entry}") from exc
+        return valid_no_proxy
+
+    @field_validator("intercept_ports_list")
+    def _validate_and_merge_ports(cls, ports: List[str]) -> List[str]:
+        """Validate and merge intercept_ports into a list of port ranges as strings."""
+        if not ports:
+            return []
+
+        ports = [p.strip() for p in ports if p.strip()]
+        if len(ports) == 1 and ports[0].upper() == "ALL":
+            return ["1-65535"]
+
+        return cls._merge_port_ranges(ports)
+
+    # ---------------- Helpers ----------------
+
+    @classmethod
+    def _merge_port_ranges(cls, ports: List[str]) -> List[str]:
+        """Merge overlapping port ranges."""
+        ranges = cls._convert_ports_to_ranges(ports)
+
+        # Merge overlapping ranges
+        ranges.sort()
+        merged_port: List[List[int]] = []
+        for start, end in ranges:
+            if not merged_port or merged_port[-1][1] < start - 1:
+                merged_port.append([start, end])
+            else:
+                merged_port[-1][1] = max(merged_port[-1][1], end)
+
+        return [f"{start}-{end}" if start != end else str(start) for start, end in merged_port]
+
+    @classmethod
+    def _convert_ports_to_ranges(cls, ports: List[str]) -> List[tuple]:
+        """Convert all port values to sorted ranges."""
+        ranges: List[tuple] = []
+        for item in ports:
+            if "-" not in item:
+                item = f"{item}-{item}"
+            try:
+                start, end = map(int, item.split("-", 1))
+            except ValueError as exc:
+                raise ValueError(f"invalid port range: {item}") from exc
+            if not 1 <= start <= end <= 65535:
+                raise ValueError(f"port range must be between 1 and 65535 instead of {item}")
+            ranges.append((start, end))
+
+        return ranges
+
+    @classmethod
+    def _get_principal_proxy_address(cls) -> str:
+        """Get proxy address from Juju-provided environment variables.
+
+        Juju automatically exposes proxy configuration to charms as environment
+        variables (JUJU_CHARM_HTTP_PROXY, JUJU_CHARM_HTTPS_PROXY).
+
+        Returns:
+            The proxy address (hostname or IP) if set, else an empty string.
+        """
+        https_proxy = os.environ.get("JUJU_HTTPS_PROXY", "")
+        http_proxy = os.environ.get("JUJU_HTTP_PROXY", "")
+
+        proxy_conf = https_proxy or http_proxy or ""
+
+        # Strip any leading http:// or https://
+        return re.sub(r"^https?://", "", proxy_conf)
```

Comment:
If the proxy string does not start with `http` or `https`, for example, `socks5h://`, then this function will return that full URL.

Reply:
Alright I've updated for it to strip any ...:// pattern